### PR TITLE
emacs: set 29 as default version and remove 28

### PR DESCRIPTION
**Pull request by jian-lin**
## Description of changes

follow-up of https://github.com/NixOS/nixpkgs/pull/247357
related https://github.com/NixOS/nixpkgs/issues/247627

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jian-lin** [comment]
Result of `nixpkgs-review pr 270558` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages marked as broken and skipped:</summary>
  <ul>
    <li>cedille</li>
    <li>emacsPackages.cedille</li>
    <li>emacsPackages.eterm-256color</li>
    <li>emacsPackages.instapaper</li>
    <li>emacsPackages.magit-stgit</li>
    <li>emacsPackages.melancholy-theme</li>
    <li>emacsPackages.nemerle</li>
    <li>emacsPackages.per-buffer-theme</li>
    <li>emacsPackages.perl-completion</li>
    <li>emacsPackages.rcirc-menu</li>
    <li>emacsPackages.shm</li>
    <li>emacsPackages.sql-presto</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>auctex</li>
    <li>auctex.tex</li>
    <li>emacsPackages.anki-editor-view</li>
    <li>emacsPackages.emacspeak</li>
    <li>emacsPackages.fixed-page-mode</li>
    <li>emacsPackages.flycheck-cfn</li>
    <li>emacsPackages.ox-leanpub</li>
  </ul>
</details>
<details>
  <summary>6067 packages built:</summary>
</details>

**Comment by jian-lin** [comment]
About build failure on x86_64-linux:
- AUCTeX 13.2 can[1] be built using Emacs 29.  However, bumping to that version causes a new build error.  Help is welcome since I am not familiar with AUCTeX.
- `emacspeak` can be built using Emacs 29 since 57.0.  See https://github.com/NixOS/nixpkgs/pull/270555.
- `anki-editor-view`, `fixed-page-mode`, `flycheck-cfn` and `ox-leanpub` fail to build because of wrong hash of `src`.  These are not caused by this PR.

[1]: https://git.savannah.gnu.org/cgit/auctex.git/commit/?id=19a2607b423f0a0eeb7530e3123cb3b70ccacd6f

**Comment by slotThe** [comment]
>     * AUCTeX 13.2 can[1](https://git.savannah.gnu.org/cgit/auctex.git/commit/?id=19a2607b423f0a0eeb7530e3123cb3b70ccacd6f) be built using Emacs 29.  However, bumping to that version causes a new build error.  Help is welcome since I am not familiar with AUCTeX.  Note that it is `pkgs.auctex`, not `pkgs.emacs.pkgs.auctex`.  I am not sure if it is necessary to have two separate auctex packages.

FWIW, the following works for me:

``` diff
diff --git a/pkgs/tools/typesetting/tex/auctex/default.nix b/pkgs/tools/typesetting/tex/auctex/default.nix
index b0fcc3b4d476..bd46955578d4 100644
--- a/pkgs/tools/typesetting/tex/auctex/default.nix
+++ b/pkgs/tools/typesetting/tex/auctex/default.nix
@@ -4,14 +4,14 @@ let auctex = stdenv.mkDerivation ( rec {
   # Make this a valid tex(live-new) package;
   # the pkgs attribute is provided with a hack below.
   pname = "auctex";
-  version = "12.3";
+  version = "13.2";
   tlType = "run";

   outputs = [ "out" "tex" ];

   src = fetchurl {
     url = "mirror://gnu/${pname}/${pname}-${version}.tar.gz";
-    hash = "sha256-L9T+MLaUV8knf+IE0+g8hHK89QDI/kqBDXREBhdMqd0=";
+    hash = "sha256-Hn5AKrz4RmlOuncZklvwlcI+8zpeZgIgHHS2ymCUQDU=";
   };

   buildInputs = [
@@ -22,6 +22,7 @@ let auctex = stdenv.mkDerivation ( rec {

   preConfigure = ''
     mkdir -p "$tex"
+    export HOME=$TMPDIR
   '';

   configureFlags = [

```

File: pkgs/top-level/aliases.nix
Comment (jian-lin) [inline pkgs/top-level/aliases.nix:220]: What should we do with these versioned aliases?

I dislike the idea that bumping these versioned every time the default Emacs version is bumped because this will keep these aliases forever.

**Comment by jian-lin** [comment]
Thanks @slotThe. I created https://github.com/NixOS/nixpkgs/pull/270810.

File: pkgs/applications/editors/emacs/default.nix
Comment (AndersonTorres) [inline pkgs/applications/editors/emacs/default.nix:35]: You are not merely setting 29 as default, you are effectively removing 28.

Put this on the description too.

File: pkgs/top-level/aliases.nix
Comment (AndersonTorres) [reply pkgs/top-level/aliases.nix:220]: They are here just because we were maintaining two versions at the same time.

There is no need to bump them.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Review by Atemu — CHANGES_REQUESTED** [review]
I don't think we should be removing 28 already. Let's do that closer to next release.

While we should obviously not carry around old baggage forever, having the two most recent Emacs versions at once is not a huge burden. 28 certainly doesn't have a rotten smell yet if you know what I mean.

The 29 by default bump should have probably already happened. Whoops. cc @figsoda @RaitoBezarius do you think we can smuggle that into the release still or is it too late?

This needs a release note either way.

**Comment by jian-lin** [comment]
> I don't think we should be removing 28 already. Let's do that closer to next release.

Could you provide your reason behind this?

The convention of Nixpkgs is to keep only one version for each package if possible.  We used to treat Emacs like that.  The fact that we have two versions of Emacs in Nixpkgs is a workaround for many broken elisp packages when we bump Emacs from 28 to 29.

In this PR, I want to fix broken elisp packages with Emacs 29 and set 29 as the default version.  After doing so,  I see no benefit to keep Emacs 28 around and no harm to remove it.

> This needs a release note either way.

Release notes are not added for the bumps of older versions like 27 and 28.  Also, we already have one release note about Emacs 29 for 23.11.

> The 29 by default bump should have probably already happened. Whoops. cc figsoda RaitoBezarius do you think we can smuggle that into the release still or is it too late?

I do not think it matters.  We can always backport changes to stable releases, right?

**Comment by Atemu** [comment]
> Could you provide your reason behind this?

There's always a little something broken. Emacs is a *huge* ecosystem, so that's pretty much a given. I think it's better to give the affected users a bit more time on this.

Additionally, we already have the infrastructure to support many versions of Emacs. We could technically probably go as low as 26 or 27 without any trouble whatsoever but that'd be a bit extreme. It's simply not a big problem to keep 28 around for a little while.

> I do not think it matters. We can always backport changes to stable releases, right?

Not this one as it updates the *default* version to a new major release.

It's not unheard of but certainly something that'd need approval of the release managers.

**Review by AndersonTorres — APPROVED** [review]

**Comment by jian-lin** [comment]
> There's always a little something broken.  Emacs is a huge ecosystem, so that's pretty much a given.

Defaulting to 29 does not break more packages than 28 after https://github.com/NixOS/nixpkgs/pull/270810 and https://github.com/NixOS/nixpkgs/pull/270555 are merged.

I do not think we should consider packages not in Nixpkgs.

> We could technically probably go as low as 26 or 27 without any trouble whatsoever

Maintenance burden is a thing.

> I think it's better to give the affected users a bit more time on this.

While it is good for affected users, it increases the maintenance burden.

As a trade-off, I propose that we only backport the commit of changing default Emacs version after we merge these two commits.

**Comment by Atemu** [comment]
The problem is that the upgrade to the default breaks a user's system. If they have some weird custom init.el that depends on features deprecated or changed in 29, that breaks their setup. We don't usually backport stuff like that unless it's absolutely necessary.

As I said, Emacs is not just some dependency of elisp packages in Nixpkgs, it's a whole ecosystem.

It's like updating GNOME in the middle of the release. We don't do that either.

> Maintenance burden is a thing.

As mentioned, I don't think it's in any way significant.

Should it become an actual burden to someone any time soon, they can remove it at that point.

Removal should absolutely not happen in stable. We shipped with it and it should stay available in stable.

**Comment by AndersonTorres** [comment]
> We could technically probably go as low as 26 or 27 without any trouble whatsoever but that'd be a bit extreme. 

The heisenbug that hindered Macport to use anything newer than LLVM 6 is a huge argument against this necro-maintenance.

**Comment by Atemu** [comment]
> > We could technically probably go as low as 26 or 27 without any trouble whatsoever but that'd be a bit extreme.
> 
> The heisenbug that hindered Macport to use anything newer than LLVM 6 is a huge argument against this necro-maintenance.

That's an entirely different circumstance; not at all comparable to keeping an emacs version around that's just a year old.  
Also, it was the newest version of macport at that time that somehow only worked with LLVM6's code generation, not some old version we kept around for "necro-maintenance". It was simply a regular plain old bug.  
If anything, this is an argument *for* keeping old versions around because we wouldn't have been able to ship macport at all if llvm6 wasn't there for all that time. (Though I would certainly not advocate for keeping old versions around this long, even some as core and interchangable as llvm.)

Even if emacs 28 had depended on some ancient version of some core library, we would simply drop it as soon as someone came around wanting to remove said library. Let's cross that bridge when we get there.

I too won't advocate for keeping every version of Emacs around forever but, in emacs terms, 29 is still quite new and 28 not very old. It's not an ecosystem that moves particularly fast (quite the opposite).

**Comment by AndersonTorres** [comment]
My suggestion would  be split  this in two commits, one changing the default and other removing the oldies.

Better granularity for future bisections.

**Comment by jian-lin** [comment]
> split this in two commits, one changing the default and other removing the oldies.

It is exactally the current state. 😅

**Comment by AndersonTorres** [comment]
I am for removing emacs 28 without backporting that removal. Next stable release it will be dead.

**Review by Atemu — APPROVED** [review]
Alright, I asked in the Nix Emacs Matrix channel and so far, nobody has brought forth any concerns regarding requiring emacs28.

I'm fine with removing emacs28 in that case, though I'd still prefer to keep it for a release.

**Comment by vcunat** [comment]
40k rebuilds on nixpkgs master?  That's too much, right?  It's certainly not just `emacsPackages`.

**Comment by Atemu** [comment]
Hm, the [eval report](https://gist.githubusercontent.com/GrahamcOfBorg/2d43f0b3dfd7e4cc6f0048633be217ba/raw/ed77912105e73ee66d81d9c7f5edc3d705dbb038/Changed%2520Paths) only lists emacsPackages.

Where are you seeing 40k packages? Is that one arch or multiple?

**Comment by vcunat** [comment]
All platforms.  And I see I counted previous eval as well.  So "only" over 36k new ones over a few hours of nixpkgs master, but maybe several different pull requests added up in there: https://hydra.nixos.org/eval/1802673

**Comment by vcunat** [comment]
I mean, maybe Hydra is slow, but I think adding 40k builds like this (then replicated in staging-next, etc.) will delay the work by at least one full day.

**Comment by vcunat** [comment]
Even with cheap builds its speed is 1-2 build steps per second over longer term.

**Comment by mweinelt** [comment]
Reverted in #272785. Please resubmit into staging.

**Comment by jian-lin** [comment]
Created https://github.com/NixOS/nixpkgs/pull/272822 which retargets this to staging.

**Comment by ckiee** [comment]
I know I'm coming in very late here, but the last working build of [nix-doom-emacs](https://github.com/nix-community/nix-doom-emacs) is on Emacs 28 and 29 [does not build unfortunately](https://github.com/nix-community/nix-doom-emacs/actions/runs/7033338605/job/19138941874). I depend on this for my day-to-day, some others probably do too though that number is dwindling. 

I'm not sure if this _should_ be a blocker, but there it is. 

(For me personally I guess I'd just revert in my fork or overlay and it's whatever. But it does break the already very broken NDE even more.)

**Comment by jian-lin** [comment]
For keeping the old 28, a natural question is how long it should be kept.

**Comment by vcunat** [comment]
I posted 1-2 build steps per second max.  24k on trunk, all again on staging-next, and half of it cached to trunk-combined.  That's more than half a day on my napkin.

**Comment by Atemu** [comment]
What exactly do you mean by "build steps"? Is that just the emacsPackages or also including the 16k unrelated packages?

**Comment by vcunat** [comment]
It's one `/nix/store/*.drv`.

**Comment by Atemu** [comment]
>> Is that just the emacsPackages or also including the 16k unrelated packages?

**Comment by vcunat** [comment]
So, simple example: 24k with master and staging-next, one full build per second = over 13 hours.
> 24000*2/3600

**Comment by Atemu** [comment]
> So, simple example: 24k with master and staging-next, one full build per second = over 13 hours.
> 
> > 24000*2/3600

I'd just like to know whether the one build/s metric was just for the emacsPackages or whether it was for the entire 40k builds that included 16k non-emacs packages. I'm asking because I'd expect those to take 10-100x as long as your typical emacs package would, thus massively skewing the average.  
For example, if it took 3h to build all 24k emacsPackages and 10h to build the 16k other rebuilds, that'd still be 13h in total but the cause wouldn't really be the emacsPackages in that hypothetical scenario.

I find it kind of hard to believe that, with all its parallelism, the entire build farm takes about as long to build four platforms as my dog slow NAS CPU needs to build one platform.  
Since separate platforms are parallel to build by definition, that must mean that one platform must have been progressing at the speed of a 10W Celeron. That doesn't seem right to me.

***

@ckiee I'd recommend you to contribute to the discussion on the new PR.

**Comment by vcunat** [comment]
1-2 build steps per second are empiric maximum over long term.  For any kind of builds apparently.  (i.e. I've seen no period longer than an hour or two where Hydra exceeded that)

**Comment by vcunat** [comment]
You know, the central place of Hydra is in Germany, but the S3 is in USA and builders are often in Finland I think or other non-Germany...

**Comment by Atemu** [comment]
> I can't have numbers for hydra.nixos.org just for emacs packages. We'd have to stop building everything else and start building just those (for some time).

Hm, would it be possible to do a rebuild of all emacs packages (i.e. no-op change in emacsPackages builder) when there's relatively little else going on?  
That way we could take end time - start time and divide by the amount of builds to get a rather good idea how many emacs package builds/s hydra can do.

**Comment by vcunat** [comment]
Theoretically yes, but it's again like... commits are waiting in `staging*` for a couple weeks and I should tell those people that they need to wait some more because I won't trigger those builds yet?

**Comment by vcunat** [comment]
So possible, but at least right now I'm not perfectly comfortable with it.

**Comment by mweinelt** [comment]
I think h.n.o is pretty busy these days through keeping up with the release branches and staging cycles.

There are regularly bigger PRs that get tested on hydra, but their feedback cycles are a few days long.

I'm running python-updates on my private infrastructure, because I can't get enough feedback from h.n.o fast enough to finish it in a meaningful timeframe.

**Comment by Atemu** [comment]
Oh, this wasn't a "we need to know this right now" request. As I said, this would necessarily be run when there's little else going on, so if there's anything else in need to be built, it's probably a higher priority than this.

**Pull request merged**
